### PR TITLE
[fixes #26] Implement LMS & NLMS adaptive filter

### LINKED ISSUE #26: Feature request: Implement LMS adaptive filtering + other adaptive filters
Adaptive filtering is often used for signal noise reduction and some other cool DSP applications. The most often used one is the Least Mean Square (LMS) adaptive filter.

A very good adaptive filtering library in Python is [padasip](https://github.com/matousc89/padasip). It also has some code documentation on how the algorithm should work, both [in code](https://github.com/matousc89/padasip/blob/master/padasip/filters/lms.py) and on a [documentation page](https://matousc89.github.io/padasip/sources/filters/lms.html).

Other potential good source: [explanation of MATLAB LMS adaptive filtering implementation](https://onlinelibrary.wiley.com/doi/pdf/10.1002/9780470745977.app2)

### PR DESCRIPTION
**Pull request by SiboVG**
This PR implements the LMS and NLMS adaptive filter.

@psambit9791 in the light of how padasip implements its functions, I used a 'run()'-method for running the filtering algorithm. This got me thinking whether it would be an idea to use this terminology in the other classes as well? So instead of having to call the 'dft()'-method on a DiscreteFourier-object, you would just have to call the 'run()'-method. In this way, the user only has to "remember" one method, i.e. the run-method, instead of knowing 'I have to run 'dft()' for a DiscreteFourier, stft() for ShortTimeFourier etc. I think this could generalize and simplify jDSP. What is your opinion on this?

### CODE REVIEW

**Comment by psambit9791** [comment]
So, regarding your comment about using 'run()', the latest commit on master converts all "dft()" "hilbert()" etc to simply "transform()" based on the sklearn implementations of some similar functions.

So at the moment, building any object of the transform package and calling the transform() function using the object would run the specific transform. Chose transform() instead of run() because it seemed more intuitive.

**Comment by SiboVG** [comment]
Ah alright, wasn't aware of that change, great :) you wanna change all the methods for the filters to 'filter()' then

**Comment by psambit9791** [comment]
That would make sense. But think it should be limited to the adaptive filters for now. The IIR filters have different initialisation depending on the type of filter so I would let them be as they are.

File: src/main/java/com/github/psambit9791/jdsp/filter/adaptive/LMSFilter.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * <h1>Least-mean-squares (LMS) adaptive filter</h1>
+ * The LMS adaptive filter is a filter that adapts its filter weights to get an input signal x to match a desired output
+ * signal (= the output of the filter). It does this by trying to minimize the squared error between the desired signal
+ * and the filter output signal.
+ *
+ * @author SiboVanGool
+ * @version 1.0
+ */
+public class LMSFilter {
```
Comment (psambit9791) [inline src/main/java/com/github/psambit9791/jdsp/filter/adaptive/LMSFilter.java:14]: Just call this LMS, conforming to the classes in ```transform``` and ```filter``` package.

File: src/main/java/com/github/psambit9791/jdsp/filter/adaptive/LMSFilter.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+import java.util.Arrays;
+
+/**
+ * <h1>Least-mean-squares (LMS) adaptive filter</h1>
+ * The LMS adaptive filter is a filter that adapts its filter weights to get an input signal x to match a desired output
+ * signal (= the output of the filter). It does this by trying to minimize the squared error between the desired signal
+ * and the filter output signal.
+ *
+ * @author SiboVanGool
```
Comment (psambit9791) [inline src/main/java/com/github/psambit9791/jdsp/filter/adaptive/LMSFilter.java:11]: Keep the name consistent between files. I think I saw "Sibo Van Gool" (with spaces) in another file.

**Comment by SiboVG** [comment]
Alright, latest commit fixed it

File: src/main/java/com/github/psambit9791/jdsp/filter/adaptive/NLMSFilter.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * <h1>Normalized Least-mean-squares (NLMS) adaptive filter</h1>
+ * The NLMS adaptive filter is a filter that adapts its filter weights to get an input signal x to match a desired output
+ * signal (= the output of the filter). It does this by trying to minimize the squared error between the desired signal
+ * and the filter output signal.
+ *
+ * It is very similar to the LMS-filter, with the difference that the learning rate gets automatically adjusted according
+ * to the input signal's power.
+ *
+ * @author SiboVanGool
```
Comment (psambit9791) [inline src/main/java/com/github/psambit9791/jdsp/filter/adaptive/NLMSFilter.java:14]: Same as above

File: src/main/java/com/github/psambit9791/jdsp/filter/adaptive/NLMSFilter.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+ * @version 1.0
+ */
+public class NLMSFilter {
```
Comment (psambit9791) [inline src/main/java/com/github/psambit9791/jdsp/filter/adaptive/NLMSFilter.java:17]: Same as above

**Review by psambit9791 — CHANGES_REQUESTED** [review]

**Pull request merged**